crates/p2p/src/swarm/dial_manager.rs

```diff
@@ -0,0 +1,79 @@
+//! Implementation of DialManager
+
+use std::{
+    collections::{HashMap, VecDeque},
+    sync::Arc,
+};
+
+use libp2p::{Multiaddr, identity::PublicKey, swarm::ConnectionId};
+use tokio::sync::Mutex;
+
+/// Manages the state for multi-address dial attempts, including address queues and connection ID
+/// mappings. Used to coordinate retries and track which connection corresponds to which dial
+/// sequence.
+#[derive(Debug, Default)]
+pub struct DialManager {
+    /// Maps an app_public_key to the queue of remaining addresses to try.
+    pub dial_queues: Arc<Mutex<HashMap<PublicKey, VecDeque<Multiaddr>>>>,
+    /// Maps a libp2p connection ID to the corresponding app_public_key.
+    pub conn_to_dial: Arc<Mutex<HashMap<ConnectionId, PublicKey>>>,
+}
+
+impl DialManager {
+    /// Creates a new, empty DialManager.
+    pub fn new() -> Self {
+        DialManager {
+            dial_queues: Arc::new(Mutex::new(HashMap::new())),
+            conn_to_dial: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+
+    /// Insert a new queue for a dial sequence id
+    pub async fn insert_queue(&self, app_public_key: PublicKey, addresses: VecDeque<Multiaddr>) {
+        let mut queues = self.dial_queues.lock().await;
+        queues.insert(app_public_key, addresses);
+    }
+
+    /// Pop the next address from the queue for a given app_public_key
+    pub async fn pop_next_addr(&self, app_public_key: &PublicKey) -> Option<Multiaddr> {
+        let mut queues = self.dial_queues.lock().await;
+        if let Some(queue) = queues.get_mut(app_public_key) {
+            queue.pop_front()
+        } else {
+            None
+        }
```

Comment:
this extends lock hold time unnecessarily during conditional logic.

When you use a lock in an if/else condition, the lock guard's lifetime extends to the entire if/else statement, not just the condition: see https://stackoverflow.com/a/79621802.